pkg/daemon/server/service/runtime/runtime.go

```diff
@@ -0,0 +1,199 @@
+	Data       []ErrorDetails `json:"data"`
+}
+type ErrorDetails struct {
+	Container string `json:"container"`
+	Timestamp string `json:"timestamp"`
+	Code      string `json:"code"`
+	Message   string `json:"message"`
+	Details   string `json:"details"`
+}
+
+type monitorHttpClient interface {
+	Get(url string) (*http.Response, error)
+	Head(url string) (*http.Response, error)
+}
+
+// pipelineRuntimeCache is used to store the local cache of runtime errors for a vertex.
+// It implements the PipelineRuntimeCache interface.
+type pipelineRuntimeCache struct {
+	pipeline   *v1alpha1.Pipeline
+	localCache map[PodReplica][]ErrorDetails
+	cacheMutex sync.RWMutex
+	podTracker *PodTracker
+	log        *zap.SugaredLogger
+	httpClient monitorHttpClient
+}
+
+func NewRuntime(ctx context.Context, pl *v1alpha1.Pipeline) PipelineRuntimeCache {
+	return &pipelineRuntimeCache{
+		pipeline:   pl,
+		localCache: make(map[PodReplica][]ErrorDetails),
+		cacheMutex: sync.RWMutex{},
+		log:        logging.FromContext(ctx).Named("pipelineRuntimeCache"),
+		httpClient: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+			Timeout: time.Second * 1,
+		},
+		podTracker: NewPodTracker(ctx, pl),
+	}
+}
+
+// StartCacheRefresher starts the cache refresher to update the local cache periodically with the runtime errors.
+func (r *pipelineRuntimeCache) StartCacheRefresher(ctx context.Context) (err error) {
+	r.log.Infof("Starting runtime server...")
+
+	ctx, cancel := context.WithCancel(logging.WithLogger(ctx, r.log))
+
+	go func() {
+		err := r.podTracker.Start(ctx)
+		if err != nil {
+			r.log.Errorw("Failed to start pod tracker for runtime server", zap.Error(err))
+			cancel()
+		}
+	}()
+
+	// start persisting errors into the local cache
+	go r.persistRuntimeErrors(ctx)
+
+	return nil
+}
+
+// persistRuntimeErrors updates the local cache with the runtime errors
+func (r *pipelineRuntimeCache) persistRuntimeErrors(ctx context.Context) {
+	// Define a function to fetch and persist errors
+	fetchAndPersistErrors := func() {
+		var wg sync.WaitGroup
+		for _, vtx := range r.pipeline.Spec.Vertices {
+			for i := range r.podTracker.GetActivePodsCountForVertex(vtx.Name) {
+				wg.Add(1)
+				go r.fetchAndPersistErrorForPod(vtx.Name, i, &wg)
+			}
+		}
+		wg.Wait()
+	}
+
+	// invoke once and then periodically update the cache
+	fetchAndPersistErrors()
+	// Set up a ticker to run periodically
+	ticker := time.NewTicker(runtimeErrorsTimeStep)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			fetchAndPersistErrors()
+		// If the context is done, return.
+		case <-ctx.Done():
+			r.log.Info("Context canceled, stopping PersistRuntimeErrors")
+			return
+		}
+	}
+}
+
+// fetchAndPersistErrorForPod fetches the runtime errors for a pod and persists them in the local cache.
+func (r *pipelineRuntimeCache) fetchAndPersistErrorForPod(vtxName string, podIndex int, wg *sync.WaitGroup) {
```

Comment:
This is risky to pass a WaitGroup as a parameter. Instead of abstracting all of these into a function, can we just do the fetching part and put the waitgroup logic back to the function `persistRuntimeErrors`?